### PR TITLE
Add LK Losses: Direct Acceptance Rate Optimization for Speculative Decoding

### PR DESCRIPTION
**Pull request by MrShevan**
## Motivation

This PR integrates the LK objectives from the paper [https://arxiv.org/pdf/2602.23881](https://arxiv.org/pdf/2602.23881) into Eagle3 training so the optimization target better matches speculative decoding behavior in practice.  
It introduces acceptance rate as an explicit optimization objective and a tracked metric during both training and evaluation, fixes the Eagle3 `acc` metric under truncated draft vocab mapping, and adds support for the Infinity-Instruct dataset used in our experiments/ablations.  

## Modifications

- Added LK-loss CLI controls in `scripts/train_eagle3.py`:
  - `--lk-loss-type` with `lambda` (hybrid KL+LK) and `alpha` (log-acceptance) modes.
  - `--kl-scale` and `--kl-decay` for adaptive KL weighting in `lambda` mode.
  - Sanity checks for non-negative KL hyperparameters.
- Added `specforge/core/lk_loss.py` with:
  - expected acceptance-rate computation,
  - masked/distributed acceptance-rate aggregation,
  - LK loss composition for `lambda` and `alpha` objectives.
- Integrated LK loss and acceptance-rate computation into Eagle3 training paths in `specforge/core/eagle3.py` (standard and Qwen-VL), with adapter state updates in `specforge/core/eagle3_adapters.py`.
- Added acceptance-rate tracking to training/eval logging in `scripts/train_eagle3.py` (`acceptance_rate_{i}` per TTT position), which is reported to the configured tracker (including W&B when enabled).
- Fixed accuracy computation in Eagle3 to compare mapped draft predictions in target-vocab space (`d2t`) against target token IDs from full target logits, avoiding errors from truncated-vocab argmax comparison.
- Added LK utility tests in `tests/test_utils/test_lk_loss_utils.py`.
- Added dataset preparation support for `nebius-llama31-8b-infinity-instruct` in `scripts/prepare_data.py`, plus `run_prepare_data.sh` convenience entrypoint.

## Related Issues

[#485](https://github.com/sgl-project/SpecForge/issues/485)

## Accuracy Test

Passed tests: `python -m unittest discover -s ./tests -p "test_*.py" -v`
Screenshot 2026-03-07 at 19 38 09

## Benchmark & Profiling
Wandb training:
Screenshot 2026-03-09 at 14 15 02
Screenshot 2026-03-09 at 14 15 59

Benchmarking Performance with a vLLM Script - [examples/offline_inference/spec_decode.py](https://github.com/vllm-project/vllm/blob/main/examples/offline_inference/spec_decode.py):
`--temp {0;1} --dataset-path philschmid/mt-bench --enable-chunked-prefill --num-prompts 80`

LK-lambda loss checkpoint (6 epochs, 120000 steps, 32 batch size, [nebius-infinity-instruct dataset](https://huggingface.co/datasets/nebius/Llama-3.1-8B-Instruct-Infinity-Instruct-0625)):
```
(Temperature: 0)
total_num_output_tokens: 16913
num_drafts: 4463
num_draft_tokens: 31241
num_accepted_tokens: 12524
mean acceptance length: 3.81
--------------------------------------------------
acceptance at token 0: 0.77
acceptance at token 1: 0.58
acceptance at token 2: 0.44
acceptance at token 3: 0.35
acceptance at token 4: 0.27
acceptance at token 5: 0.22
acceptance at token 6: 0.18

(Temperature: 1)
total_num_output_tokens: 17172
num_drafts: 4890
num_draft_tokens: 34230
num_accepted_tokens: 12358
mean acceptance length: 3.53
--------------------------------------------------
acceptance at token 0: 0.73
acceptance at token 1: 0.54
acceptance at token 2: 0.40
acceptance at token 3: 0.30
acceptance at token 4: 0.23
acceptance at token 5: 0.18
acceptance at token 6: 0.14
```

LK-alpha loss checkpoint (6 epochs, 120000 steps, 32 batch size, [nebius-infinity-instruct dataset](https://huggingface.co/datasets/nebius/Llama-3.1-8B-Instruct-Infinity-Instruct-0625)):
```
(Temperature: 0)
total_num_output_tokens: 17030
num_drafts: 4500
num_draft_tokens: 31500
num_accepted_tokens: 12592
mean acceptance length: 3.80
--------------------------------------------------
acceptance at token 0: 0.77
acceptance at token 1: 0.58
acceptance at token 2: 0.44
acceptance at token 3: 0.35
acceptance at token 4: 0.27
acceptance at token 5: 0.22
acceptance at token 6: 0.18

(Temperature: 1)
total_num_output_tokens: 17220
num_drafts: 4983
num_draft_tokens: 34881
num_accepted_tokens: 12290
mean acceptance length: 3.47
--------------------------------------------------
acceptance at token 0: 0.73
acceptance at token 1: 0.53
acceptance at token 2: 0.39
acceptance at token 3: 0.30
acceptance at token 4: 0.22
acceptance at token 5: 0.17
acceptance at token 6: 0.13
```

KL loss checkpoint (6 epochs, 120000, 32 batch size, [nebius-infinity-instruct dataset](https://huggingface.co/datasets/nebius/Llama-3.1-8B-Instruct-Infinity-Instruct-0625))
```
(Temperature: 0)
total_num_output_tokens: 17030
num_drafts: 4589
num_draft_tokens: 32123
num_accepted_tokens: 12501
mean acceptance length: 3.72
--------------------------------------------------
acceptance at token 0: 0.77
acceptance at token 1: 0.58
acceptance at token 2: 0.44
acceptance at token 3: 0.33
acceptance at token 4: 0.25
acceptance at token 5: 0.20
acceptance at token 6: 0.15

(Temperature: 1)
total_num_output_tokens: 17183
num_drafts: 5212
num_draft_tokens: 36484
num_accepted_tokens: 12025
mean acceptance length: 3.31
--------------------------------------------------
acceptance at token 0: 0.73
acceptance at token 1: 0.52
acceptance at token 2: 0.37
acceptance at token 3: 0.26
acceptance at token 4: 0.19
acceptance at token 5: 0.14
acceptance at token 6: 0.10
```

Reference model: `yuhuili/EAGLE3-LLaMA3.1-Instruct-8B`
```
(Temperature: 0)
total_num_output_tokens: 16934
num_drafts: 4942
num_draft_tokens: 34594
num_accepted_tokens: 12024
mean acceptance length: 3.43
--------------------------------------------------
acceptance at token 0: 0.75
acceptance at token 1: 0.53
acceptance at token 2: 0.39
acceptance at token 3: 0.28
acceptance at token 4: 0.21
acceptance at token 5: 0.16
acceptance at token 6: 0.12
```

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).

### CODE REVIEW

**Comment by astralord** [comment]
Hi, @shuaills @sleepcoo @FlamingoPg @FrankLeeeee,

Can you please clarify, if you plan to review/merge this PR?

Would appreciate any feedback,
Thanks!

**Comment by FrankLeeeee** [comment]
@astralord thanks for your contribution, sorry for outlooking this. I will take the review job.

**Comment by astralord** [comment]
Hi, @FrankLeeeee, any updates on this?

**Comment by FrankLeeeee** [comment]
Hi  @astralord , I am running experiments for this PR, should be done today

**Comment by MrShevan** [comment]
Hi @FrankLeeeee, just checking whether there’s anything else I can help with here — for example, additional benchmarks or follow-up changes — if that would make the review easier. Thanks!

**Comment by jiapingW** [comment]
Sorry, I'll check it now.
